csharp/ql/lib/semmle/code/csharp/security/dataflow/ZipSlipQuery.qll

```diff
@@ -3,147 +3,360 @@
  */
 
 import csharp
-private import semmle.code.csharp.controlflow.Guards
+import semmle.code.csharp.dataflow.TaintTracking2
+import semmle.code.csharp.dataflow.TaintTracking3
+import semmle.code.csharp.dataflow.TaintTracking4
+import semmle.code.csharp.controlflow.Guards
+import semmle.code.csharp.dataflow.DataFlow::DataFlow::PathGraph
 
-/**
- * A data flow source for unsafe zip extraction.
- */
-abstract class Source extends DataFlow::Node { }
+abstract class AbstractSanitizerMethod extends Method { }
 
-/**
- * A data flow sink for unsafe zip extraction.
- */
-abstract class Sink extends DataFlow::ExprNode { }
+class RootSanitizerMethod extends AbstractSanitizerMethod {
+  // Alias for String.StartsWith
+  RootSanitizerMethod() { this.getQualifiedName() = "System.String.StartsWith" }
+}
 
-/**
- * A sanitizer for unsafe zip extraction.
- */
-abstract class Sanitizer extends DataFlow::ExprNode { }
+abstract class UnsanitizedPathCombiner extends Expr { }
 
-/**
- * DEPRECATED: Use `Sanitizer` instead.
- *
- * A guard for unsafe zip extraction.
- */
-abstract deprecated class SanitizerGuard extends DataFlow::BarrierGuard { }
+class PathCombinerViaMethodCall extends UnsanitizedPathCombiner {
+  PathCombinerViaMethodCall() {
+    exists(MethodCall mcPathCombine |
+      mcPathCombine.getTarget().hasQualifiedName("System.IO.Path", "Combine") and
+      this = mcPathCombine
+    )
+  }
+}
 
-/** A taint tracking configuration for Zip Slip */
-class TaintTrackingConfiguration extends TaintTracking::Configuration {
-  TaintTrackingConfiguration() { this = "ZipSlipTaintTracking" }
+class PathCombinerViaStringInterpolation extends UnsanitizedPathCombiner {
+  PathCombinerViaStringInterpolation() { exists(InterpolatedStringExpr e | this = e) }
```

Comment:
Is there a parent to this that is a general concat?

Reply:
TODO: Look at superclass of InterpolatedStringExpr

Consider moving Path combination classes out to a generic library

Reply:
Answer is No